numba/np/ufunc/gufunc.py

```diff
@@ -0,0 +1,116 @@
+        self.gufunc_builder = GUFuncBuilder(
+            py_func, signature, identity, cache, targetoptions)
+
+    def add(self, fty):
+        self.gufunc_builder.add(fty)
+
+    def build_ufunc(self):
+        self.ufunc = self.gufunc_builder.build_ufunc()
+        return self
+
+    @property
+    def __doc__(self):
+        return self.ufunc.__doc__
+
+    @property
+    def __name__(self):
+        return self.ufunc.__name__
+
+    @property
+    def nin(self):
+        return self.ufunc.nin
+
+    @property
+    def nout(self):
+        return self.ufunc.nout
+
+    @property
+    def nargs(self):
+        return self.ufunc.nargs
+
+    @property
+    def ntypes(self):
+        return self.ufunc.ntypes
+
+    @property
+    def types(self):
+        return self.ufunc.types
+
+    @property
+    def identity(self):
+        return self.ufunc.identity
+
+    def _get_argtys_dtype(self, args):
+        argtys = map(lambda x: typeof(x), args)
+        tys = []
+        for argty in argtys:
+            if isinstance(argty, types.Array):
+                tys.append(argty.dtype)
+            else:
+                tys.append(argty)
+        return tys
+
+    def _get_element_wise_signature(self, *args):
+        parsed_sig = parse_signature(self.signature)
+        dtypes = self._get_argtys_dtype(args)  # [int32, int32, int32, ...]
+
+        # Two cases here can happen here
+        # 1. args contains the output array:
+        #      gufunc(A, B, C)
+        # 2. args doesn't contains the output array:
+        #      C = gufunc(A, B)
+        #
+        # In the latter case, one would have to guess the type
+        # and format (scalar or array). This behavior will be
+        # forbidden for now.
+        # See: https://github.com/numba/numba/pull/5938#issuecomment-661978921
+
+        # self.signature[1] has always length 1
+        if len(dtypes) < len(self.signature[0]) + 1:
+            msg = '`out = gufunc(Arg1, Arg2, ..., ArgN)` is not allowed.'
+            '\nUse `gufunc(Arg1, Arg2, ..., ArgN, out)` instead'
+            raise TypeError(msg)
+
+        # first time calling the gufunc
+        # generate a signature based on input arguments
+        l = []
+        for idx, sig_dim in enumerate(parsed_sig[0]):
+            ndim = len(sig_dim)
+            if ndim == 0:  # append scalar
+                l.append(dtypes[idx])
+            else:
+                l.append(types.Array(dtypes[idx], ndim, 'A'))
+
+        # add return to signature
+        retty = dtypes[-1]
+        ret_ndim = len(parsed_sig[-1][0]) or 1  # small hack to return scalar
+        l.append(types.Array(retty, ret_ndim, 'A'))
+
+        return types.none(*l)
+
+    def __call__(self, *args, **kwargs):
+        try:
+            return self.ufunc(*args, **kwargs)
+        except TypeError:
+            sig = self._get_element_wise_signature(*args)
+            self.add(sig)
+            self.build_ufunc()
+            return self.ufunc(*args, **kwargs)
```

Comment:
Is there a better way to detect if a gufunc was already compiled for the given types?

Reply:
Is it ok to check if a gufunc was already compiled in the way I did? Or should I follow the same approach as dufunc?

Reply:
I originally thought it was ok. However, after looking into the message in `TypeError`, the exception is probably too generic. First, we can't tell if the error is `TypeError("'NoneType' object is not callable")` because `self.ufunc` is `None`. Secondly, we can't tell if the error is indeed caused by problem with type coercion. Can you try to use the `ufunc_find_matching_loop` so that it can be more precise about what is happening?